src/libraries/System.Private.CoreLib/src/System/Environment.OSVersion.OSX.cs

```diff
@@ -7,7 +7,15 @@ public static partial class Environment
     {
         private static OperatingSystem GetOSVersion()
         {
-            Version version = Interop.libobjc.GetOperatingSystemVersion();
```

Comment:
What was wrong with the existing solution?

Reply:
(And if there was something wrong with it, why is it still ok for mono?)

Reply:
The main issues are around the how the DllImport to objc_msgSend isn't really accurate (@AaronRobinsonMSFT can explain more his feeling behind this) and how we have to manually determine which message passing function (objc_msgSend or objc_msgSend_stret) depending on target platform when the rules are not obvious and the native compiler will handle it for us transparently. Additionally, if we're already going through the work to enable Objective-C in the build system for the Objective-C interop, we can reduce the chattiness of multiple interop calls here by moving it to native and making it one call.

The goal is to move mono over as well in the future (or do this in the libraries layer as @jkotas mentioned) and remove the `Interop.libobjc` class in its current form.

Reply:
I would interesting in details what makes this specific instance of `objc_msgSend` P/Invoke wrong. I do not see any problem with it.

> we can reduce the chattiness of multiple interop calls here by moving it to native and making it one call.

We prefer to write in C# where possible. I think we should resort to rewriting C# in C++ only where it makes measurable performance difference.

Reply:
It isn't wrong from a technical perspective, since that is how it can be done. However, what is stated is misleading and isn't what is really desired . This API shouldn't be considered an export of `objc_msgSend` because it isn't. In this form it is a specialized use case for `objc_msgSend` with no arguments and that is confusing given how this function actually operates. We also hit concerns with one of the new APIs being introduced around overriding the export look up. I was preferring this to be done in Objective-C naturally instead of expose its weirdness at the C# level.

This also makes the C# far more common for those who maintain it and pushes the Objective-C semantics to be the responsibility of the Objective-C compiler.

Reply:
If we want to do this, I think it should be in separate PR, target main branch, and done for both CoreCLR and Mono at the same time.

Reply:
I did a quick search around the runtime and I saw that after this PR, the System.Drawing.Common library is the last place where we P/Invoke into any `objc_msgSend`-like functions. We should take a look there as well at some point to clean up their usage of the methods to get the function pointers from the new APIs.

Reply:
Ok I'll move this into a different branch.

Reply:
> We should take a look there as well at some point to clean up their usage of the methods to get the function pointers from the new APIs.

I do not see why we would want to do it. This is doing lean one-off objective-C interop and I think it is perfectly fine. I do not think we should be introducing dependencies on the full objective-C interop unnecessarily.

Reply:
> I do not think we should be introducing dependencies on the full objective-C interop unnecessarily.

Okay. That is a fair argument. The idea was to move all Objective-C logic into an Objective-C native library but if the dependency chain isn't desirable lets scrap it.

@jkoritzinsky Can you at least rename the `objc_sendMsg` to something that indicates what it is doing rather than the generic name?

Reply:
I'll do that in another PR targeting master as mentioned.

Reply:
What would you rename it to? The naming convention for interop under libraries is to use the name that the platform headers or documentation uses. `objc_sendMsg` is exactly that name in this case.

Reply:
Opened https://github.com/dotnet/runtime/pull/47608